### PR TITLE
WL-3959:  Add Section button

### PR DESCRIPTION
**Pull request by nicholaswilson100**
It may be that this ORLIMS stuff needs
its own js file in future as there's quite a lot of it and I'll create one when we get to that stage.

### CODE REVIEW

**Comment by nicholaswilson100** [comment]
The db stuff in WL-3961 (https://github.com/ox-it/wl-citations/pull/20) should probably be merged before this code which relies on it.

File: citations-impl/impl/src/java/org/sakaiproject/citation/impl/BasicCitationService.java [hunk 1]
```diff
@@ -264,6 +265,22 @@ public void saveCollection(CitationCollection collection)
         }
 
 		/* (non-Javadoc)
+         * @see org.sakaiproject.citation.impl.BaseCitationService.Storage#saveCitationCollectionOrder(org.sakaiproject.citation.api.CitationCollectionOrder)
```
Comment (buckett) [inline citations-impl/impl/src/java/org/sakaiproject/citation/impl/BasicCitationService.java]: Indentation looks wrong here.

File: citations-api/api/src/java/org/sakaiproject/citation/api/CitationCollectionOrder.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
+package org.sakaiproject.citation.api;
+
+/**
+ * Created by nickwilson on 6/23/15.
+ */
+public class CitationCollectionOrder {
+
+	private String collectionId;
+	private int location;
+	private String sectionType;
```
Comment (buckett) [inline citations-api/api/src/java/org/sakaiproject/citation/api/CitationCollectionOrder.java]: How about an enum?

File: citations-tool/tool/src/java/org/sakaiproject/citation/tool/CitationHelperAction.java [hunk 3]
```diff
@@ -979,6 +972,27 @@ protected void doPostJsonResponse(ParameterParser params,
...
+			String addSectionHTML = params.getString("addSectionHTML");
+			int locationId = params.getInt("locationId");
+			CitationCollection collection = getCitationCollection(state, false);
+			CitationCollectionOrder citationCollectionOrder = new CitationCollectionOrder(collection.getId(), locationId, "heading1", addSectionHTML);
+			getCitationService().save(citationCollectionOrder);
+			message = "Resource updated";
+		}
+		catch (Exception e){
+			message = e.getMessage();
+			logger.warn("Exception in addSection() " + e);
```
Comment (buckett) [inline citations-tool/tool/src/java/org/sakaiproject/citation/tool/CitationHelperAction.java:994]: It's normally better to log the full exception `log.warn(String, Exception)` if it's unexpected as it's something that going to need to be debugged, if it's expected then it shouldn't be logged at warn level.

Comment (buckett) [inline citations-tool/tool/src/java/org/sakaiproject/citation/tool/CitationHelperAction.java:994]: Ah, I see it's the existing pattern in the code.

File: citations-tool/tool/src/java/org/sakaiproject/citation/tool/CitationHelperAction.java [hunk 4]
```diff
@@ -979,6 +972,27 @@ protected void doPostJsonResponse(ParameterParser params,
 		return results;
 	}
 
+	protected Map<String, Object> addSection(ParameterParser params, SessionState state) {
+		String message;
+		Map<String, Object> results = new HashMap<String, Object>();
+		try {
+			String addSectionHTML = params.getString("addSectionHTML");
```
Comment (buckett) [inline citations-tool/tool/src/java/org/sakaiproject/citation/tool/CitationHelperAction.java:983]: You need to add some filtering for bad HTML so that you can't put evil script tags in.

File: citations-tool/tool/src/java/org/sakaiproject/citation/tool/CitationHelperAction.java [hunk 5]
```diff
@@ -979,6 +972,27 @@ protected void doPostJsonResponse(ParameterParser params,
...
+			int locationId = params.getInt("locationId");
+			CitationCollection collection = getCitationCollection(state, false);
+			CitationCollectionOrder citationCollectionOrder = new CitationCollectionOrder(collection.getId(), locationId, "heading1", addSectionHTML);
+			getCitationService().save(citationCollectionOrder);
+			message = "Resource updated";
```
Comment (buckett) [inline citations-tool/tool/src/java/org/sakaiproject/citation/tool/CitationHelperAction.java]: If you're displaying this to the user (which I think you are), this needs i18n.

File: citations-tool/tool/src/webapp/js/new_resource.js [hunk 6]
```diff
@@ -786,5 +792,93 @@ citations_new_resource.init = function() {
...
+        CKEDITOR.instances[this.id].on( 'blur', function( evt ) {
+
+            var actionUrl = $('#newCitationListForm').attr('action');
+            $('#citation_action').val('add_section');
+            var params = $('#newCitationListForm').serializeArray();
+            params.push({name:'addSectionHTML', value:$(evt.editor.getData()).html()});
+            params.push({name:'locationId', value:evt.editor.name.substring('sectionInlineEditor'.length, evt.editor.name.length)});
+
+            $.ajax({
+                type		: 'POST',
```
Comment (buckett) [inline citations-tool/tool/src/webapp/js/new_resource.js]: Indenting inside the line with tabs leads to uneven indentation when the tab size is different.

File: citations-tool/tool/src/webapp/js/new_resource.js [hunk 7]
```diff
@@ -786,5 +792,93 @@ citations_new_resource.init = function() {
...
+                url			: actionUrl,
+                cache		: false,
+                data		: params,
+                dataType	: 'json',
+                success		: function(jsObj) {
```
Comment (buckett) [inline citations-tool/tool/src/webapp/js/new_resource.js]: Looks like it should be extracted out as a method as the other added success function is the same.

File: citations-tool/tool/src/webapp/js/new_resource.js [hunk 8]
```diff
@@ -786,5 +792,93 @@ citations_new_resource.init = function() {
...
+
+    });
+
+    // show inline ckeditor for section
+    $('#addSectionButton').on('click', function(e) {
+
+        var locationId = citations_new_resource.getNextLocationId();
+        citations_new_resource.locationId++;
+        var divId = 'sectionInlineEditor' + locationId;
+        var html = "<div id='" + divId + "' title='Click here to edit' contenteditable='true' style='border: 1px white solid'>" +
```
Comment (buckett) [inline citations-tool/tool/src/webapp/js/new_resource.js]: i18n here again.

File: citations-tool/tool/src/webapp/js/new_resource.js [hunk 9]
```diff
@@ -786,5 +792,93 @@ citations_new_resource.init = function() {
...
+            "<div class='listTitle' style='background-color: #002147; color:#FFF;'><h1>Section Title</h1></div></div>";
```
Comment (buckett) [inline citations-tool/tool/src/webapp/js/new_resource.js]: i18n here again.

**Comment by buckett** [comment]
In the long run I'd try and make the Citations API more useful so it can return a document with the elements within it already sorted and ordered as a tree of items (citations, headings, etc). At the moment it's rather close the just an abstraction of SQL.

File: citations-tool/tool/src/webapp/vm/citation/new_resource.vm [hunk 10]
```diff
@@ -441,6 +441,20 @@ span.searchTargetLink { margin-right:50px; }
...
+					#set( $divId = 'sectionInlineEditor' + $num )
+					<div id='$divId' title='Click here to edit' contenteditable='true' style='border: 1px white solid'>
+						<div class='listTitle' style='background-color: #002147; color:#FFF;'>
+							$nestedSection.getValue()
+						</div>
+					</div>
+				#end
+			#end
+			<div class="act" id="addSectionDiv">
+				<input id="addSectionButton" type="button" name="AddSection" class="active" value="Add Section"/>
```
Comment (buckett) [inline citations-tool/tool/src/webapp/vm/citation/new_resource.vm]: i18n here

File: citations-tool/tool/src/webapp/vm/citation/new_resource.vm [hunk 11]
```diff
@@ -441,6 +441,20 @@ span.searchTargetLink { margin-right:50px; }
 			<div class="act">
 				<input type="button" name="Cancel" class="Cancel active" value="$tlang.getString("resource.done")" />
 			</div>
+			#if($nestedSectionsSize && $nestedSectionsSize > 0)
+				#foreach($nestedSection in $nestedSections)
+					#set( $num = $velocityCount - 1)
+					#set( $divId = 'sectionInlineEditor' + $num )
+					<div id='$divId' title='Click here to edit' contenteditable='true' style='border: 1px white solid'>
```
Comment (buckett) [inline citations-tool/tool/src/webapp/vm/citation/new_resource.vm]: i18n here

File: citations-api/api/src/java/org/sakaiproject/citation/api/CitationCollectionOrder.java (hunk 2 above)
Comment (nicholaswilson100) [inline citations-api/api/src/java/org/sakaiproject/citation/api/CitationCollectionOrder.java]: In what kind of way?  Enums are lists of constants whereas this class is a POJO that (for example) gets populated dynamcially from a row from CITATION_COLLECTION _ORDER.

File: citations-tool/tool/src/java/org/sakaiproject/citation/tool/CitationHelperAction.java (hunk 4 above)
Comment (nicholaswilson100) [inline citations-tool/tool/src/java/org/sakaiproject/citation/tool/CitationHelperAction.java:983]: This needs doing before we release but is it okay to put it in a different pull request as I just want to get the basic editing functionality working first and then fix security issues like this?

File: citations-impl/impl/src/java/org/sakaiproject/citation/impl/BasicCitationService.java (hunk 1 above)
Comment (nicholaswilson100) [inline citations-impl/impl/src/java/org/sakaiproject/citation/impl/BasicCitationService.java]: ok

File: citations-tool/tool/src/java/org/sakaiproject/citation/tool/CitationHelperAction.java (hunk 5 above)
Comment (nicholaswilson100) [inline citations-tool/tool/src/java/org/sakaiproject/citation/tool/CitationHelperAction.java]: ok

File: citations-tool/tool/src/webapp/js/new_resource.js (hunk 6 above)
Comment (nicholaswilson100) [inline citations-tool/tool/src/webapp/js/new_resource.js]: ok

File: citations-tool/tool/src/webapp/js/new_resource.js (hunk 7 above)
Comment (nicholaswilson100) [inline citations-tool/tool/src/webapp/js/new_resource.js]: ok

**Comment by nicholaswilson100** [comment]
Re: In the long run I'd try and make the Citations API more useful so it can return a document with the elements within it already sorted and ordered as a tree of items (citations, headings, etc). At the moment it's rather close the just an abstraction of SQL.

Agreed.  I was just getting it working basically and building iteratively on that.  I can see eventually the API needing to be more flexible and  constructing something that is tree like that represents more closely a nested reading list.

File: citations-tool/tool/src/webapp/js/new_resource.js (hunk 8 above)
Comment (nicholaswilson100) [inline citations-tool/tool/src/webapp/js/new_resource.js]: bit trickier in the js...hmm

Comment (nicholaswilson100) [inline citations-tool/tool/src/webapp/js/new_resource.js]: I could put something on the page that the js grabs when it loads I suppose

Comment (nicholaswilson100) [inline citations-tool/tool/src/webapp/js/new_resource.js]: not great

File: citations-tool/tool/src/webapp/js/new_resource.js (hunk 9 above)
Comment (nicholaswilson100) [inline citations-tool/tool/src/webapp/js/new_resource.js]: I can actually remove this as it's not needed for now

File: citations-tool/tool/src/webapp/js/new_resource.js (hunk 8 above)
Comment (nicholaswilson100) [inline citations-tool/tool/src/webapp/js/new_resource.js]: I can actually remove this for now as it's not needed (but there's the same js/i18n problem below)

File: citations-api/api/src/java/org/sakaiproject/citation/api/CitationCollectionOrder.java (hunk 2 above)
Comment (buckett) [inline citations-api/api/src/java/org/sakaiproject/citation/api/CitationCollectionOrder.java]: But the sectionType can only be a limited number of options so I'd use an enum to make the API easier to use.

Comment (nicholaswilson100) [inline citations-api/api/src/java/org/sakaiproject/citation/api/CitationCollectionOrder.java]: Oh I see yup will do

File: citations-tool/tool/src/webapp/js/new_resource.js (hunk 8 above)
Comment (buckett) [inline citations-tool/tool/src/webapp/js/new_resource.js]: Have a hidden element in the page that is copied in, or use a templating library (if you're going that way in the long run)?

Comment (nicholaswilson100) [inline citations-tool/tool/src/webapp/js/new_resource.js]: yeah

**Comment by nicholaswilson100** [comment]
The various fixes above include fixing indentation, SectionType as enum, making strings i18n, and factoring out ajax call.

**Comment by nicholaswilson100** [comment]
Comment on previous commit: I've added a red cross image to the top right of a section (silk/cross.png) which the user can click and it will delete the section from the db and, if the ajax post is successful, removes it from the page also.

File: citations-impl/impl/src/java/org/sakaiproject/citation/impl/BasicCitationService.java [hunk 12]
```diff
@@ -263,7 +264,30 @@ public void saveCollection(CitationCollection collection)
         	this.m_collections.put(collection.getId(), collection);
         }
 
+        /* (non-Javadoc)
+         * @see org.sakaiproject.citation.impl.BaseCitationService.Storage#saveCitationCollectionOrder(org.sakaiproject.citation.api.CitationCollectionOrder)
+         */
+		public void saveCitationCollectionOrder(CitationCollectionOrder citationCollectionOrder)
```
Comment (buckett) [inline citations-impl/impl/src/java/org/sakaiproject/citation/impl/BasicCitationService.java:261]: Indentation.

File: citations-impl/impl/src/java/org/sakaiproject/citation/impl/DbCitationService.java [hunk 13]
```diff
@@ -323,6 +335,76 @@ else if(value instanceof String)
         }
 
 		/* (non-Javadoc)
+         * @see org.sakaiproject.citation.impl.BaseCitationService.Storage#commitCitationCollectionOrder(org.sakaiproject.citation.api.CitationCollectionOrder)
```
Comment (buckett) [inline citations-impl/impl/src/java/org/sakaiproject/citation/impl/DbCitationService.java]: Indentation.

File: citations-impl/impl/src/java/org/sakaiproject/citation/impl/BasicCitationService.java [hunk 14]
```diff
@@ -263,7 +264,30 @@ public void saveCollection(CitationCollection collection)
...
+		{
+
+		}
+
+		/* (non-Javadoc)
+		* @see org.sakaiproject.citation.impl.BaseCitationService.Storage#getNestedSections(java.lang.String)
+		*/
+		public List<CitationCollectionOrder> getNestedSections(String citationCollectionId)
```
Comment (buckett) [inline citations-impl/impl/src/java/org/sakaiproject/citation/impl/BasicCitationService.java:269]: Why have this method at all?

File: citations-impl/impl/src/java/org/sakaiproject/citation/impl/BasicCitationService.java [hunk 15]
```diff
@@ -263,7 +264,30 @@ public void saveCollection(CitationCollection collection)
...
+		{
+			return null;
+		}
 		/* (non-Javadoc)
+		* @see org.sakaiproject.citation.impl.BaseCitationService.Storage#remove(java.lang.String, int, org.sakaiproject.citation.api.CitationCollectionOrder.SectionType)
+		*/
+		public void remove (String collectionId, int locationId, CitationCollectionOrder.SectionType sectionType)
+		{
+
```
Comment (buckett) [inline citations-impl/impl/src/java/org/sakaiproject/citation/impl/BasicCitationService.java]: Why have this method at all?

File: citations-impl/impl/src/java/org/sakaiproject/citation/impl/DbCitationService.java [hunk 16]
```diff
@@ -1377,6 +1459,40 @@ public Object readSqlResultRecord(ResultSet result)
...
+			String sectionType = null;
+			String value = null;
+			try
+			{
+				collectionId = result.getString(1);
+				location = result.getInt(2);
+				sectionType = result.getString(3);
+				value = result.getString(4);
+
+				citationCollectionOrder = new CitationCollectionOrder(collectionId, location, CitationCollectionOrder.SectionType.HEADING1, value);
```
Comment (buckett) [inline citations-impl/impl/src/java/org/sakaiproject/citation/impl/DbCitationService.java:1483]: Shouldn't you be parsing this to an enum?

File: citations-tool/tool/src/java/org/sakaiproject/citation/tool/CitationHelperAction.java (hunk 4 above)
Comment (buckett) [inline citations-tool/tool/src/java/org/sakaiproject/citation/tool/CitationHelperAction.java:983]: That's a good way to have it missed for the release, why would we want to release this with a know security hole?

**Comment by nicholaswilson100** [comment]
Fixed as comments above (up to last commit)

**Comment by nicholaswilson100** [comment]
Rebased all commits so far

**Comment by nicholaswilson100** [comment]
Still need to do HTML cleaning and a couple of other tidies....

**Comment by nicholaswilson100** [comment]
HTML-cleaning of a section
